lib/auth/requestable_roles_test.go

```diff
@@ -0,0 +1,263 @@
+		}
+		require.Empty(t, cmp.Diff(expectedRequestableRoles[3:6], receivedRoles))
+
+		// Verify there is no overlap in roles between the pages.
+		firstPageRoleNames := make(map[string]bool)
+		for _, role := range firstPageResp.Roles {
+			firstPageRoleNames[role.GetName()] = true
+		}
+		for _, role := range secondPageResp.Roles {
+			require.False(t, firstPageRoleNames[role.GetName()])
+		}
+	})
+
+	t.Run("list all pages of requestable roles", func(t *testing.T) {
+		limit := int32(3)
+		var respRoles []*types.RoleV6
+		nextKey := ""
+
+		for {
+			req := &authpb.ListRolesRequest{
+				Limit:    limit,
+				StartKey: nextKey,
+				Filter: &types.RoleFilter{
+					RequestableOnly: true,
+				},
+			}
+
+			resp, err := a.ListRoles(userCtx, req)
+			require.NoError(t, err)
+
+			respRoles = append(respRoles, resp.Roles...)
+
+			if resp.NextKey == "" {
+				break
+			}
+
+			// Verify that we got the correct page size.
+			if resp.NextKey != "" {
+				require.Len(t, resp.Roles, int(limit))
+			}
+
+			nextKey = resp.NextKey
+		}
+
+		// Verify that all the requestable roles were returned.
+		var receivedRoles []string
+		for _, role := range respRoles {
+			receivedRoles = append(receivedRoles, role.GetName())
+		}
+		require.Empty(t, cmp.Diff(expectedRequestableRoles, receivedRoles))
+	})
+
+	t.Run("no requestable roles", func(t *testing.T) {
+		// Create a user with no requestable roles
+		noPermsUsername := "noperms-user"
+		noPermsRole, err := types.NewRole("noperms-role", types.RoleSpecV6{
+			Allow: types.RoleConditions{
+				Logins: []string{"ubuntu"},
+			},
+			// No request permissions.
+		})
+		require.NoError(t, err)
+
+		_, err = a.CreateRole(ctx, noPermsRole)
+		require.NoError(t, err)
+
+		noPermsUser, err := types.NewUser(noPermsUsername)
+		require.NoError(t, err)
+		noPermsUser.SetRoles([]string{"noperms-role"})
+
+		_, err = a.CreateUser(ctx, noPermsUser)
+		require.NoError(t, err)
+
+		noPermsIdentity := tlsca.Identity{
+			Username: noPermsUsername,
+			Groups:   []string{"noperms-role"},
+		}
+		restrictedCtx := authz.ContextWithUser(ctx, authz.LocalUser{
+			Username: noPermsUsername,
+			Identity: noPermsIdentity,
+		})
+
+		req := &authpb.ListRolesRequest{
+			Filter: &types.RoleFilter{
+				RequestableOnly: true,
+			},
+		}
+		resp, err := a.ListRoles(restrictedCtx, req)
+		require.NoError(t, err)
+
+		// Verify that nothing is returned.
+		require.Empty(t, resp.Roles)
+		require.Empty(t, resp.NextKey)
+	})
+
+	t.Run("list requestable roles with a search filter", func(t *testing.T) {
+		req := &authpb.ListRolesRequest{
+			Filter: &types.RoleFilter{
+				SearchKeywords:  []string{"role-99"},
+				RequestableOnly: true,
+			},
+		}
+		resp, err := a.ListRoles(userCtx, req)
+		require.NoError(t, err)
+
+		// Verify that only "role-99" was returned
+		require.Len(t, resp.Roles, 1)
+		require.Equal(t, resp.Roles[0].GetName(), "role-99")
+
+		// There shouldn't be a nextKey
+		require.Empty(t, resp.NextKey)
+	})
+
```

Comment:
Is there a chance that a deny directive was not evaluated because the role is not available in the current page?  
Can that be tested? 

Sorry if I am not making any sense, just trying to wrap my head around it.

Reply:
Hmm could you clarify with an example? I don't understand what you mean